### PR TITLE
upgrade to use sdk v0.11.0

### PR DESCRIPTION
**Pull request by camilamacedo86**
**Description**
- Upgrade the project to use SDK 0.11
- Fix CSV file

### CODE REVIEW

File: memcached-operator/pkg/apis/cache/v1alpha1/memcached_types.go [hunk 1]
```diff
@@ -24,15 +25,27 @@ type MemcachedStatus struct {
...
 	// Add custom validation using kubebuilder tags: https://book-v1.book.kubebuilder.io/beyond_basics/generating_crd.html
 
-	// Status of Nodes
-	Nodes []string `json:"nodes"`
+	// Nodes are the names of the memcached pods
+	// +kubebuilder:validation:UniqueItems=true
+	Nodes PodList `json:"nodes"`
+}
+
+//PodList is a list of namespaces
```
Comment (camilamacedo86) [inline memcached-operator/pkg/apis/cache/v1alpha1/memcached_types.go]:
```suggestion
//PodList is a list of podNames
```

File: memcached-operator/.gitignore [hunk 2]
```diff
@@ -78,3 +75,4 @@ tags
 .vscode/*
 .history
 # End of https://www.gitignore.io/api/go,vim,emacs,visualstudiocode
+.idea/
```
Comment (joelanford) [inline memcached-operator/.gitignore]: Is this change necessary? Looks like you just moved the `.idea` entry?

File: memcached-operator/deploy/olm-catalog/memcached-operator/0.0.1/memcached-operator.v0.0.1.clusterserviceversion.yaml [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    - memcached
+    - app
+  maintainers:
+    - email: corp@example.com
+      name: Some Corp
+  maturity: alpha
+  provider:
+    name: Example
+    url: www.example.com
+  version: 0.0.1
```
Comment (joelanford) [inline memcached-operator/deploy/olm-catalog/memcached-operator/0.0.1/memcached-operator.v0.0.1.clusterserviceversion.yaml]: Is there a reason we deleted the existing CSVs and created a new one at a lower semver version? I would think we should just create a new CSV at `v0.10.1` or `v0.11.0` depending on whether there are breaking changes.

File: memcached-operator/pkg/apis/cache/v1alpha1/memcached_types.go [hunk 4]
```diff
@@ -24,15 +25,27 @@ type MemcachedStatus struct {
...
+}
+
+// PodList is a list of names of pods
+type PodList struct {
+	Names []string
+}
+
+func NewPodList(value []string) PodList {
+	p := PodList{Names: value}
+	return p
```
Comment (joelanford) [inline memcached-operator/pkg/apis/cache/v1alpha1/memcached_types.go]: I'm not sure I understand the need for nesting this. Is this a structural schema thing? I'd be pretty surprised if structural schema forbids top-level `[]string` values. 

Is there a doc that talks about this?

File: memcached-operator/deploy/olm-catalog/memcached-operator/0.0.1/memcached-operator.v0.0.1.clusterserviceversion.yaml (hunk 3 above)
Comment (camilamacedo86) [reply memcached-operator/deploy/olm-catalog/memcached-operator/0.0.1/memcached-operator.v0.0.1.clusterserviceversion.yaml]: - The previous files were just generated with operator-sdk but was not updated then they never actually worked. What is the point to keep them? But if you wish, I can come back them to the project.
- Regards the version, see that we are using the tag 0.0.1 so to keep aligned with it. Note that the tags used before shows that were matching with the SDK releases and not the breaking changes into it. 

IMHO, we could apply the fix since now this CSV will work and start to upgrade/change it based on the changes in the project. WDYT? Keep it as it is now in this PR or came back and just add the 0.11.0 on it so?

File: memcached-operator/pkg/apis/cache/v1alpha1/memcached_types.go (hunk 4 above)
Comment (camilamacedo86) [reply memcached-operator/pkg/apis/cache/v1alpha1/memcached_types.go]: You can easily test it and check the error in the Validations rules. ( just clone the project, add []string and run `operator-sdk generate openapi`).  Following the error:

```
 $ operator-sdk generate openapi
INFO[0000] Running OpenAPI code-generation for Custom Resource group versions: [cache:[v1alpha1], ] 
API rule violation: list_type_missing,github.com/operator-framework/operator-sdk-samples/memcached-operator/pkg/apis/cache/v1alpha1,MemcachedStatus,Nodes
INFO[0006] Created deploy/crds/cache.example.com_memcacheds_crd.yaml 
INFO[0006] Code-generation complete.  
```
Also, see https://github.com/operator-framework/operator-sdk/issues/2114 have a better understanding and check why we did not see it before. 

Also, I working on to try to identify if is an issue in the openapi, if it will be solved in the feature versions and etc .. However .. we cannot keep the example, guide, mock data etc with something that will not work. Also, the Type allow the users to know how they can do that which shows a common question as well.

File: memcached-operator/.gitignore (hunk 2 above)
Comment (camilamacedo86) [reply memcached-operator/.gitignore]: I will remove it.

File: memcached-operator/pkg/apis/cache/v1alpha1/memcached_types.go (hunk 4 above)
Comment (joelanford) [reply memcached-operator/pkg/apis/cache/v1alpha1/memcached_types.go]: Sounds like maybe we just need an extra annotation that we didn't need before?

Comment (camilamacedo86) [reply memcached-operator/pkg/apis/cache/v1alpha1/memcached_types.go]: Let's with who shows be the master on this. @estroz, do you know another way to solve this scenario? Do you think that it can be some annotation?

Comment (camilamacedo86) [reply memcached-operator/pkg/apis/cache/v1alpha1/memcached_types.go]: Raised : https://github.com/kubernetes/kube-openapi/issues/175

Comment (camilamacedo86) [reply memcached-operator/pkg/apis/cache/v1alpha1/memcached_types.go]: confirmed the []string is no longer acceptable 👍

**Comment by camilamacedo86** [comment]
Hi @joelanford, 

So, confirmed with the k8s maintainers the []string is no longer acceptable and the reverted the change in the gitignore. So, let me know if you are ok with and if we can merge this one?

**Comment by camilamacedo86** [comment]
Hi @joelanford, 

I think that all are addressed here. Could you please let us know if you are ok with now and if we can merge this one?

File: memcached-operator/pkg/apis/cache/v1alpha1/memcached_types.go [hunk 5]
```diff
@@ -24,7 +25,9 @@ type MemcachedStatus struct {
 	// Important: Run "operator-sdk generate k8s" to regenerate code after modifying this file
 	// Add custom validation using kubebuilder tags: https://book-v1.book.kubebuilder.io/beyond_basics/generating_crd.html
 
-	// Status of Nodes
+	// Nodes are the names of the memcached pods
+	// +kubebuilder:validation:UniqueItems=true
+	// +listType=set  //More Info: https://godoc.org/k8s.io/kube-openapi/pkg/idl
```
Comment (joelanford) [inline memcached-operator/pkg/apis/cache/v1alpha1/memcached_types.go]: Move the `//More info` comment, as described here: https://github.com/operator-framework/operator-sdk/pull/2130#discussion_r340929945

File: memcached-operator/deploy/olm-catalog/memcached-operator/0.0.1/memcached-operator.v0.0.1.clusterserviceversion.yaml (hunk 3 above)
Comment (estroz) [reply memcached-operator/deploy/olm-catalog/memcached-operator/0.0.1/memcached-operator.v0.0.1.clusterserviceversion.yaml]: The current project structure suggests that each release of the SDK has a corresponding version of memcached-operator and CSV. That CSV should therefore reference:

1. The new memcached-operator image, ex. `quay.io/example/memcached-operator:v0.y.0`
1. The previous memcached-operator version in `spec.replaces`, ex. `v0.x.0`
1. The new memcached-operator version in `spec.version`, ex. `v0.y.0`

where x is the old and y is the new minor version. The CSV's should be tested to make sure they are not broken on each release of this repo.

We might want to consider building and pushing example memcached-operator images to quay.io to maintain reproducibility.

Comment (camilamacedo86) [reply memcached-operator/deploy/olm-catalog/memcached-operator/0.0.1/memcached-operator.v0.0.1.clusterserviceversion.yaml]: After a chat with @estroz was decided came them back and keep the new one as 0.11.0

**Comment by camilamacedo86** [comment]
Hi @joelanford and @estroz, 

Feel free to check it now. All done. 
After this PR we need to upgrade it to 0.12.0 :-)

File: memcached-operator/deploy/olm-catalog/memcached-operator/memcached-operator.package.yaml [hunk 6]
```diff
@@ -1,5 +1,5 @@
 channels:
-- currentCSV: memcached-operator.v0.10.0
```
Comment (estroz) [inline memcached-operator/deploy/olm-catalog/memcached-operator/memcached-operator.package.yaml]: We should be adding each working CSV name to `channels` with name `spec.version` from the CSV. The default channel should be updated to the newest channel name too.

File: memcached-operator/deploy/olm-catalog/memcached-operator/0.11.0/memcached-operator.v0.11.0.clusterserviceversion.yaml [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+                          fieldRef:
+                            fieldPath: metadata.annotations['olm.targetNamespaces']
+                      - name: POD_NAME
+                        valueFrom:
+                          fieldRef:
+                            fieldPath: metadata.name
+                      - name: OPERATOR_NAME
+                        value: memcached-operator
+                    image: quay.io/example/memcached-operator:v0.0.1
+                    imagePullPolicy: Never
```
Comment (joelanford) [inline memcached-operator/deploy/olm-catalog/memcached-operator/0.11.0/memcached-operator.v0.11.0.clusterserviceversion.yaml]: The other CSVs in this bundle use:
```
image: REPLACE_IMAGE
imagePullPolicy: Always
```

Any reason we aren't continuing using the same convention? Thoughts on this change @estroz?

File: memcached-operator/deploy/olm-catalog/memcached-operator/0.11.0/memcached-operator.v0.11.0.clusterserviceversion.yaml [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+                    name: memcached-operator
+                    resources: {}
+                serviceAccountName: memcached-operator
+      permissions:
+        - rules:
+            - apiGroups:
+                - ""
+              resources:
+                - pods
+                - services
```
Comment (joelanford) [inline memcached-operator/deploy/olm-catalog/memcached-operator/0.11.0/memcached-operator.v0.11.0.clusterserviceversion.yaml]: The CSV for v0.10.0 includes another rule for `services/finalizers`. Is that okay to remove?

File: memcached-operator/deploy/olm-catalog/memcached-operator/0.6.0/memcached-operator.v0.6.0.clusterserviceversion.yaml [hunk 9]
```diff
@@ -1,124 +0,0 @@
...
-  - app
-  maintainers:
-  - email: corp@example.com
-    name: Some Corp
-  maturity: alpha
-  provider:
-    name: Example
-    url: www.example.com
-  replaces: memcached-operator.v0.5.0
-  version: 0.6.0
```
Comment (joelanford) [inline memcached-operator/deploy/olm-catalog/memcached-operator/0.6.0/memcached-operator.v0.6.0.clusterserviceversion.yaml]: Why are we removing this CSV and keeping the others?

File: memcached-operator/deploy/olm-catalog/memcached-operator/0.11.0/memcached-operator.v0.11.0.clusterserviceversion.yaml (hunk 7 above)
Comment (estroz) [reply memcached-operator/deploy/olm-catalog/memcached-operator/0.11.0/memcached-operator.v0.11.0.clusterserviceversion.yaml]: I'd keep what the other CSV's have. We should discuss if/how we want to build and tag memcached-operator images for this repo. IMO we should get CI set up and push a new image on each release to be referenced by the CSV.

File: memcached-operator/deploy/olm-catalog/memcached-operator/0.11.0/memcached-operator.v0.11.0.clusterserviceversion.yaml [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+                        valueFrom:
+                          fieldRef:
+                            fieldPath: metadata.annotations['olm.targetNamespaces']
+                      - name: POD_NAME
+                        valueFrom:
+                          fieldRef:
+                            fieldPath: metadata.name
+                      - name: OPERATOR_NAME
+                        value: memcached-operator
+                    image: quay.io/example/memcached-operator:v0.0.1
```
Comment (camilamacedo86) [inline memcached-operator/deploy/olm-catalog/memcached-operator/0.11.0/memcached-operator.v0.11.0.clusterserviceversion.yaml]:
```suggestion
                    image: REPLACE_IMAGE
```

File: memcached-operator/deploy/olm-catalog/memcached-operator/0.11.0/memcached-operator.v0.11.0.clusterserviceversion.yaml (hunk 7 above)
Comment (camilamacedo86) [inline memcached-operator/deploy/olm-catalog/memcached-operator/0.11.0/memcached-operator.v0.11.0.clusterserviceversion.yaml]:
```suggestion
                    imagePullPolicy: Always
```

File: memcached-operator/deploy/olm-catalog/memcached-operator/0.11.0/memcached-operator.v0.11.0.clusterserviceversion.yaml (hunk 8 above)
Comment (camilamacedo86) [reply memcached-operator/deploy/olm-catalog/memcached-operator/0.11.0/memcached-operator.v0.11.0.clusterserviceversion.yaml]: The test-framework is not using this role. However, the roles here should be the same as the role.yaml.
So, excellent good catcher. Fixed.

File: memcached-operator/deploy/olm-catalog/memcached-operator/0.11.0/memcached-operator.v0.11.0.clusterserviceversion.yaml (hunk 7 above)
Comment (camilamacedo86) [reply memcached-operator/deploy/olm-catalog/memcached-operator/0.11.0/memcached-operator.v0.11.0.clusterserviceversion.yaml]: Hi @estroz, 

It shows a great idea. But it would be out of the scope of this PR now. 
I think we need to address many improvements here.

File: memcached-operator/deploy/olm-catalog/memcached-operator/memcached-operator.package.yaml (hunk 6 above)
Comment (camilamacedo86) [reply memcached-operator/deploy/olm-catalog/memcached-operator/memcached-operator.package.yaml]: Could you give an example over how to fix it? Also, could we not do it in another PR? It shows not part of the scope to upgrade the SDK version on it.

File: memcached-operator/deploy/olm-catalog/memcached-operator/0.6.0/memcached-operator.v0.6.0.clusterserviceversion.yaml (hunk 9 above)
Comment (camilamacedo86) [reply memcached-operator/deploy/olm-catalog/memcached-operator/0.6.0/memcached-operator.v0.6.0.clusterserviceversion.yaml]: It was missing to be reverted. Good catch.

File: memcached-operator/deploy/olm-catalog/memcached-operator/memcached-operator.package.yaml [hunk 11]
```diff
@@ -1,5 +1,5 @@
...
+- currentCSV: memcached-operator.v0.11.0
```
Comment (camilamacedo86) [inline memcached-operator/deploy/olm-catalog/memcached-operator/memcached-operator.package.yaml]:
```suggestion
- currentCSV: v0.11.0
```

File: memcached-operator/deploy/olm-catalog/memcached-operator/memcached-operator.package.yaml (hunk 6 above)
Comment (camilamacedo86) [reply memcached-operator/deploy/olm-catalog/memcached-operator/memcached-operator.package.yaml]: Tks I catcher now 👍

File: memcached-operator/pkg/controller/memcached/memcached_controller.go [hunk 12]
```diff
@@ -2,15 +2,13 @@ package memcached
 
 import (
 	"context"
-	"reflect"
-
 	cachev1alpha1 "github.com/operator-framework/operator-sdk-samples/memcached-operator/pkg/apis/cache/v1alpha1"
+	"reflect"
```
Comment (joelanford) [inline memcached-operator/pkg/controller/memcached/memcached_controller.go]: Nit: This `"reflect"` change should be reverted. `cachev1alpha1` should not be grouped with other standard library imports.

Out convention is to group standard libs, third-party libs, and project libs separately.

File: memcached-operator/tools.go [hunk 13]
```diff
@@ -0,0 +1,5 @@
+// +build tools
+
+// Place any runtime dependencies as imports in this file.
+// Go modules will be forced to download and install them.
+package tools
```
Comment (joelanford) [inline memcached-operator/tools.go]: Remove this file since `memcached-operator` doesn't have any runtime dependencies?

File: memcached-operator/deploy/olm-catalog/memcached-operator/memcached-operator.package.yaml (hunk 6 above)
Comment (joelanford) [reply memcached-operator/deploy/olm-catalog/memcached-operator/memcached-operator.package.yaml]: Sorry just for the sake of my own understanding, what was the resolution to this?

File: memcached-operator/tools.go (hunk 13 above)
Comment (camilamacedo86) [reply memcached-operator/tools.go]: The scaffold will add it by default. It was not added before because it is something which was introduced in the 0.9 version. So, should we not keep it?

File: memcached-operator/pkg/controller/memcached/memcached_controller.go [hunk 14]
```diff
@@ -5,12 +5,11 @@ import (
 	"reflect"
 
 	cachev1alpha1 "github.com/operator-framework/operator-sdk-samples/memcached-operator/pkg/apis/cache/v1alpha1"
-
+
```
Comment (camilamacedo86) [inline memcached-operator/pkg/controller/memcached/memcached_controller.go]:
```suggestion
```

File: memcached-operator/tools.go (hunk 13 above)
Comment (joelanford) [reply memcached-operator/tools.go]: This is a nit, so not a merge blocker, but IMO the samples should reflect how we would actually develop the operator. If it were me developing the operator, I would probably remove this file from the project since there are not runtime dependencies. But that's my personal preference. No strong opinion on this.

Comment (camilamacedo86) [reply memcached-operator/tools.go]: Removed 👍

File: memcached-operator/deploy/olm-catalog/memcached-operator/memcached-operator.package.yaml [hunk 15]
```diff
@@ -1,5 +1,5 @@
...
   name: alpha
```
Comment (camilamacedo86) [inline memcached-operator/deploy/olm-catalog/memcached-operator/memcached-operator.package.yaml]:
```suggestion
  name: 0.11.0
```

File: memcached-operator/deploy/olm-catalog/memcached-operator/memcached-operator.package.yaml (hunk 6 above)
Comment (joelanford) [reply memcached-operator/deploy/olm-catalog/memcached-operator/memcached-operator.package.yaml]: @estroz Can you clarify your original suggestion. Are you thinking we can have multiple channels, like this:

```yaml
channels:
- currentCSV: memcached-operator.v0.11.0
  name: v0.11.x
- currentCSV: memcached-operator.v0.10.0
  name: v0.10.x
defaultChannel: v0.11.x
```

Comment (camilamacedo86) [reply memcached-operator/deploy/olm-catalog/memcached-operator/memcached-operator.package.yaml]: I spoke with @estroz and he explained that the name should be the version `0.11.0`. 

@estroz is it ok for you now?

Comment (estroz) [reply memcached-operator/deploy/olm-catalog/memcached-operator/memcached-operator.package.yaml]: The version format in the channel name doesn't matter to me. @joelanford the current change looks good?

Comment (joelanford) [reply memcached-operator/deploy/olm-catalog/memcached-operator/memcached-operator.package.yaml]: As is, I don't think I agree with this change. I think we either:
1. Stick with `alpha` and don't add any more _new_ channels.
2. OR Use `v0.11.x`, `v0.10.x`, etc. for each CSV we have that goes with a specific version of the SDK. That way if we make patch releases, we don't have to have a whole new channel for the CSV. 

To be totally honest, I think it's a little weird to have CSVs per SDK version, since that isn't how a "real" operator would do things.

My vote would be to go with option 1 since that seems to be the way we've been doing it.

Comment (camilamacedo86) [reply memcached-operator/deploy/olm-catalog/memcached-operator/memcached-operator.package.yaml]: +1 for @joelanford, @estroz are you ok with? Can I do the option 1 here?

Comment (estroz) [reply memcached-operator/deploy/olm-catalog/memcached-operator/memcached-operator.package.yaml]: @joelanford what you're implying is that we should update the `alpha` channel with the new CSV version on each release?

An issue crops up here with the CSV's `spec.replaces`. If OLM can't resolve that CSV, then installation will fail. So each `spec.replaces` must reference a CSV name specified in the package manifest. That is why I suggested we use versioned channel names. We may have to remove `spec.replaces` from all CSV versions if we do not want to also expand the number of channels.

Comment (camilamacedo86) [reply memcached-operator/deploy/olm-catalog/memcached-operator/memcached-operator.package.yaml]: Since the goal here is just to migrate to upper version, I think that we should keep it as it has been done so far and then we discuss these points to get done in another PR. Tracked https://github.com/operator-framework/operator-sdk-samples/issues/89 for it. WDYT? Could we move forward in this way? 

@joelanford @estroz
